File: frontend/apps/ios/Pantopus/Features/SupportTrains/Manage/ManageTrainViewModel.swift
//
//  ManageTrainViewModel.swift
//  Pantopus
//
//  A13.13 — Manage train. Organizer-side surface for an active support
//  train. The VM holds the editable update draft (message body, audience
//  selection, push-to-phones toggle) and the close-train sheet draft
//  (optional thank-you note + presentation state). On `Send update` we
//  clear the draft + flash a toast; on `Close & thank` we flip the
//  train to `.closed` and dismiss the sheet. No backend round-trips —
//  the projection is purely UI state.
//

import Foundation

// MARK: - Content models

/// One audience chip in the Send-an-update form (`All helpers 12` etc).
public struct AudienceChipContent: Sendable, Hashable, Identifiable {
    public let id: String
    public let label: String
    public let count: String

    public init(id: String, label: String, count: String) {
        self.id = id
        self.label = label
        self.count = count
    }
}

/// Visual tone for an Organize-section row's leading icon tile.
public enum OrganizeRowTone: Sendable, Hashable {
    case amber
    case sky
    case green
    case red
}

/// One row in the Organize section card (or the Close-train destructive row).
public struct OrganizeRowContent: Sendable, Hashable, Identifiable {
    public let id: String
    public let icon: PantopusIcon
    public let tone: OrganizeRowTone
    public let label: String
    public let meta: String?
    public let sub: String?
    public let isDestructive: Bool

    public init(
        id: String,
        icon: PantopusIcon,
        tone: OrganizeRowTone,
        label: String,
        meta: String?,
        sub: String?,
        isDestructive: Bool
    ) {
        self.id = id
        self.icon = icon
        self.tone = tone
        self.label = label
        self.meta = meta
        self.sub = sub
        self.isDestructive = isDestructive
    }
}

/// The CloseTrainSheet's static copy. Editable thank-you note lives on the VM.
public struct CloseTrainSheetContent: Sendable, Hashable {
    public let daysEarlyLabel: String
    public let mealsDelivered: String
    public let neighborsHelped: String
    public let coverageDays: String
    public let recipientQuote: String

    public init(
        daysEarlyLabel: String,
        mealsDelivered: String,
        neighborsHelped: String,
        coverageDays: String,
        recipientQuote: String
    ) {
        self.daysEarlyLabel = daysEarlyLabel
        self.mealsDelivered = mealsDelivered
        self.neighborsHelped = neighborsHelped
        self.coverageDays = coverageDays
        self.recipientQuote = recipientQuote
    }
}

/// Static, design-driven content for one Manage Train screen instance.
public struct ManageTrainContent: Sendable, Hashable {
    public let trainId: String
    public let title: String
    public let dateRangeLabel: String
    public let isActive: Bool

    // 4-cell StatCellRow values + tones (tones are derived in the view).
    public let slotFillValue: String
    public let helpersValue: String
    public let daysLeftValue: String
    public let dropoutValue: String

    // SlotPreview mini-fill strip — paints 21 dots in 3 buckets.
    public let slotsFilled: Int
    public let slotsOpen: Int
    public let slotsDropout: Int
    public let slotsTotal: Int
    public let slotFillCaption: String

    // Send-an-update form
    public let draftMessage: String
    public let audienceChips: [AudienceChipContent]
    public let selectedAudienceId: String
    public let pushToPhones: Bool

    // Organize + wind-down rows
    public let organizeRows: [OrganizeRowContent]
    public let closeRow: OrganizeRowContent

    // Close-train sheet
    public let close: CloseTrainSheetContent

    public init(
        trainId: String,
        title: String,
        dateRangeLabel: String,
        isActive: Bool,
        slotFillValue: String,
        helpersValue: String,
        daysLeftValue: String,
        dropoutValue: String,
        slotsFilled: Int,
        slotsOpen: Int,
        slotsDropout: Int,
        slotsTotal: Int,
        slotFillCaption: String,
        draftMessage: String,
        audienceChips: [AudienceChipContent],
        selectedAudienceId: String,
        pushToPhones: Bool,
        organizeRows: [OrganizeRowContent],
        closeRow: OrganizeRowContent,
        close: CloseTrainSheetContent
    ) {
        self.trainId = trainId
        self.title = title
        self.dateRangeLabel = dateRangeLabel
        self.isActive = isActive
        self.slotFillValue = slotFillValue
        self.helpersValue = helpersValue
        self.daysLeftValue = daysLeftValue
        self.dropoutValue = dropoutValue
        self.slotsFilled = slotsFilled
        self.slotsOpen = slotsOpen
        self.slotsDropout = slotsDropout
        self.slotsTotal = slotsTotal
        self.slotFillCaption = slotFillCaption
        self.draftMessage = draftMessage
        self.audienceChips = audienceChips
        self.selectedAudienceId = selectedAudienceId
        self.pushToPhones = pushToPhones
        self.organizeRows = organizeRows
        self.closeRow = closeRow
        self.close = close
    }
}

/// Aggregate UI state.
public enum ManageTrainState: Sendable, Equatable {
    case loading
    case loaded(ManageTrainContent)
    case error(message: String)
}

/// Drives the Close-train confirmation sheet.
public enum ManageTrainSheetMode: Sendable, Hashable {
    case hidden
    case closing
    case closed
}

/// Character cap for the update-message textarea. Mirrors the
/// "168/500" counter in the design source.
public let manageTrainMessageMaxChars: Int = 500

@Observable
@MainActor
public final class ManageTrainViewModel {
    public private(set) var state: ManageTrainState = .loading

    /// Editable draft fields. Initialised from `content.draftMessage` /
    /// `selectedAudienceId` / `pushToPhones` on `load()`.
    public var draftMessage: String = ""
    public var selectedAudienceId: String = ""
    public var pushToPhones: Bool = true

    /// Editable thank-you note typed inside the close sheet.
    public var thankYouNote: String = ""

    /// Close-train sheet presentation state.
    public var sheetMode: ManageTrainSheetMode = .hidden

    /// Transient toast (e.g. "Update sent · 12 helpers").
    public var toast: String?

    private let trainId: String
    private let seed: ManageTrainContent?

    public init(trainId: String, content: ManageTrainContent? = nil) {
        self.trainId = trainId
        seed = content
    }

    public func load() async {
        let content = seed ?? ManageTrainSampleData.active
        state = .loaded(content)
        draftMessage = content.draftMessage
        selectedAudienceId = content.selectedAudienceId
        pushToPhones = content.pushToPhones
        thankYouNote = ""
        sheetMode = .hidden
    }

    public func refresh() async { await load() }

    // MARK: - Send-update form

    public var characterCount: Int { draftMessage.count }
    public var characterCounterLabel: String {
        "\(characterCount) / \(manageTrainMessageMaxChars)"
    }

    /// True when the draft message has at least one non-whitespace
    /// character and is under the cap. Mirrors the design's
    /// `Send update` enable rule (textarea non-empty + valid).
    public var canSendUpdate: Bool {
        let trimmed = draftMessage.trimmingCharacters(in: .whitespacesAndNewlines)
        return !trimmed.isEmpty && characterCount <= manageTrainMessageMaxChars
    }

    public func updateDraftMessage(_ value: String) {
        // Hard-clip to the cap so the counter never displays over-limit.
        if value.count > manageTrainMessageMaxChars {
            draftMessage = String(value.prefix(manageTrainMessageMaxChars))
        } else {
            draftMessage = value
        }
    }

    public func selectAudience(_ id: String) {
        guard case let .loaded(content) = state,
              content.audienceChips.contains(where: { $0.id == id }) else { return }
        selectedAudienceId = id
    }

    public func togglePush(_ value: Bool) { pushToPhones = value }

    /// Send the typed update. Clears the draft + flashes a toast so the
    /// helper count surfaces. Real `POST /api/support-trains/:id/updates`
    /// wiring lands when the backend ships.
    public func sendUpdate() {
        guard canSendUpdate, case let .loaded(content) = state else { return }
        let helperCount = content.audienceChips.first { $0.id == selectedAudienceId }?.count
            ?? content.helpersValue
        draftMessage = ""
        toast = "Update sent · \(helperCount) helpers"
    }

    public func acknowledgeToast() { toast = nil }

    // MARK: - Close-train sheet

    public func showCloseSheet() { sheetMode = .closing }
    public func hideCloseSheet() { sheetMode = .hidden }

    public func updateThankYouNote(_ value: String) { thankYouNote = value }

    /// Flip the train to `.closed`. The sheet dismisses and the train's
    /// chip flips from "Active" green to "Closed" neutral.
    public func confirmClose() {
        guard case let .loaded(content) = state else { return }
        let next = ManageTrainContent(
            trainId: content.trainId,
            title: content.title,
            dateRangeLabel: content.dateRangeLabel,
            isActive: false,
            slotFillValue: content.slotFillValue,
            helpersValue: content.helpersValue,
            daysLeftValue: content.daysLeftValue,
            dropoutValue: content.dropoutValue,
            slotsFilled: content.slotsFilled,
            slotsOpen: content.slotsOpen,
            slotsDropout: content.slotsDropout,
            slotsTotal: content.slotsTotal,
            slotFillCaption: content.slotFillCaption,
            draftMessage: content.draftMessage,
            audienceChips: content.audienceChips,
            selectedAudienceId: content.selectedAudienceId,
            pushToPhones: content.pushToPhones,
            organizeRows: content.organizeRows,
            closeRow: content.closeRow,
            close: content.close
        )
        state = .loaded(next)
        sheetMode = .closed
        toast = "Train closed · thanks sent to \(content.helpersValue) helpers"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/support_trains/manage/ManageTrainViewModel.kt
@file:Suppress("MagicNumber", "PackageNaming")

package app.pantopus.android.ui.screens.support_trains.manage

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.support_trains.SupportTrainUpdateBody
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.support_trains.SupportTrainsRepository
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * A13.13 — Manage train. Aggregate UI state for the organizer-side
 * surface. Mirrors the iOS `ManageTrainState` shape so the two
 * platforms project identical content.
 */
sealed interface ManageTrainState {
    /** Initial fetch in flight (shimmer skeleton). */
    object Loading : ManageTrainState

    /** Train loaded successfully. */
    data class Loaded(val content: ManageTrainContent) : ManageTrainState

    /** Fetch failed; the screen surfaces the message with a `Try again` CTA. */
    data class Error(val message: String) : ManageTrainState
}

/** One audience chip in the Send-an-update form. */
data class AudienceChipContent(
    val id: String,
    val label: String,
    val count: String,
)

/** Visual tone for an Organize-section row's leading icon tile. */
enum class OrganizeRowTone { AMBER, SKY, GREEN, RED }

/** One row in the Organize section card (or the Close-train destructive row). */
data class OrganizeRowContent(
    val id: String,
    val icon: PantopusIcon,
    val tone: OrganizeRowTone,
    val label: String,
    val meta: String?,
    val sub: String?,
    val isDestructive: Boolean,
)

/** The CloseTrainSheet's static copy. The editable thank-you note lives on the VM. */
data class CloseTrainSheetContent(
    val daysEarlyLabel: String,
    val mealsDelivered: String,
    val neighborsHelped: String,
    val coverageDays: String,
    val recipientQuote: String,
)

/** Static, design-driven content for one Manage Train screen instance. */
data class ManageTrainContent(
    val trainId: String,
    val title: String,
    val dateRangeLabel: String,
    val isActive: Boolean,
    val slotFillValue: String,
    val helpersValue: String,
    val daysLeftValue: String,
    val dropoutValue: String,
    val slotsFilled: Int,
    val slotsOpen: Int,
    val slotsDropout: Int,
    val slotsTotal: Int,
    val slotFillCaption: String,
    val draftMessage: String,
    val audienceChips: List<AudienceChipContent>,
    val selectedAudienceId: String,
    val pushToPhones: Boolean,
    val organizeRows: List<OrganizeRowContent>,
    val closeRow: OrganizeRowContent,
    val close: CloseTrainSheetContent,
)

/** Drives the Close-train confirmation sheet presentation. */
enum class ManageTrainSheetMode { HIDDEN, CLOSING, CLOSED }

/**
 * Wire-format UI state: the content frame + the editable draft + the
 * sheet + the transient toast. Mirrors the iOS `ManageTrainViewModel`
 * surface so parity tests can compare projections directly.
 */
data class ManageTrainUiState(
    val state: ManageTrainState = ManageTrainState.Loading,
    val draftMessage: String = "",
    val selectedAudienceId: String = "",
    val pushToPhones: Boolean = true,
    val thankYouNote: String = "",
    val sheetMode: ManageTrainSheetMode = ManageTrainSheetMode.HIDDEN,
    val toast: String? = null,
) {
    val characterCount: Int get() = draftMessage.length
    val characterCounterLabel: String get() = "$characterCount / $MAX_MESSAGE_CHARS"

    /**
     * True when the draft message has at least one non-whitespace
     * character and is under the cap. Mirrors the design's
     * `Send update` enable rule.
     */
    val canSendUpdate: Boolean
        get() =
            draftMessage.trim().isNotEmpty() &&
                draftMessage.length <= MAX_MESSAGE_CHARS

    companion object {
        const val MAX_MESSAGE_CHARS: Int = 500
    }
}

/**
 * P4.3 — A13.13 — Manage train ViewModel.
 *
 * `load()` fetches `GET /api/support-trains/:id` and derives the organizer
 * dashboard ([ManageTrainProjection]); pass a `seed` to render offline
 * (previews / tests). `sendUpdate` posts to `POST /:id/updates`;
 * `confirmClose` marks the train completed via `POST /:id/complete`
 * (sending the optional thank-you note as a final broadcast first). Both
 * mutate local state optimistically so the toast / chip flip stay instant.
 *
 * PROJECTION GAPS (no backend field): audience segmentation + push-to-phones
 * stay client-only (the endpoint broadcasts to everyone); dropout shows 0;
 * there is no single "close with thanks" route.
 */
@HiltViewModel
class ManageTrainViewModel
    @Inject
    constructor(
        private val repo: SupportTrainsRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val trainId: String =
            savedStateHandle.get<String>(TRAIN_ID_KEY).orEmpty()

        private val _state = MutableStateFlow(ManageTrainUiState())
        val state: StateFlow<ManageTrainUiState> = _state.asStateFlow()

        /**
         * Load the dashboard. With a `seed` (previews / tests) it renders
         * directly; otherwise it fetches `GET /:id` and projects it.
         */
        fun load(seed: ManageTrainContent? = null) {
            if (seed != null) {
                applyContent(seed)
                return
            }
            _state.update { ManageTrainUiState(state = ManageTrainState.Loading) }
            viewModelScope.launch {
                when (val result = repo.detail(trainId)) {
                    is NetworkResult.Success ->
                        applyContent(ManageTrainProjection.project(result.data))
                    is NetworkResult.Failure ->
                        _state.update { it.copy(state = ManageTrainState.Error(result.error.message)) }
                }
            }
        }

        private fun applyContent(content: ManageTrainContent) {
            _state.update {
                ManageTrainUiState(
                    state = ManageTrainState.Loaded(content),
                    draftMessage = content.draftMessage,
                    selectedAudienceId = content.selectedAudienceId,
                    pushToPhones = content.pushToPhones,
                    thankYouNote = "",
                    sheetMode = ManageTrainSheetMode.HIDDEN,
                    toast = null,
                )
            }
        }

        // MARK: - Send-update form

        fun updateDraftMessage(value: String) {
            // Hard-clip to the cap so the counter never displays over-limit.
            val clamped =
                if (value.length > ManageTrainUiState.MAX_MESSAGE_CHARS) {
                    value.substring(0, ManageTrainUiState.MAX_MESSAGE_CHARS)
                } else {
                    value
                }
            _state.update { it.copy(draftMessage = clamped) }
        }

        fun selectAudience(id: String) {
            val current = _state.value
            val content = (current.state as? ManageTrainState.Loaded)?.content ?: return
            if (content.audienceChips.none { it.id == id }) return
            _state.update { it.copy(selectedAudienceId = id) }
        }

        fun togglePush(value: Boolean) {
            _state.update { it.copy(pushToPhones = value) }
        }

        /**
         * Send the typed update via `POST /api/support-trains/:id/updates`.
         * Optimistically clears the draft + flashes the toast; the audience
         * filter + push-to-phones toggle have no backend field (the endpoint
         * broadcasts to everyone) so they stay client-only.
         */
        fun sendUpdate() {
            val current = _state.value
            if (!current.canSendUpdate) return
            val content = (current.state as? ManageTrainState.Loaded)?.content ?: return
            val body = current.draftMessage
            val helperCount =
                content.audienceChips.firstOrNull { it.id == current.selectedAudienceId }?.count
                    ?: content.helpersValue
            _state.update {
                it.copy(
                    draftMessage = "",
                    toast = "Update sent · $helperCount helpers",
                )
            }
            viewModelScope.launch {
                repo.postUpdate(trainId, SupportTrainUpdateBody(body = body))
            }
        }

        fun acknowledgeToast() {
            _state.update { it.copy(toast = null) }
        }

        // MARK: - Close-train sheet

        fun showCloseSheet() {
            _state.update { it.copy(sheetMode = ManageTrainSheetMode.CLOSING) }
        }

        fun hideCloseSheet() {
            _state.update { it.copy(sheetMode = ManageTrainSheetMode.HIDDEN) }
        }

        fun updateThankYouNote(value: String) {
            _state.update { it.copy(thankYouNote = value) }
        }

        /**
         * Close & thank. Optimistically flips the train to `closed`, then
         * sends the thank-you note as a final broadcast (`POST /:id/updates`)
         * when one was typed and marks the train completed
         * (`POST /:id/complete`). The backend has no single "close with
         * thanks" route, so this composes the two calls.
         */
        fun confirmClose() {
            val current = _state.value
            val content = (current.state as? ManageTrainState.Loaded)?.content ?: return
            val note = current.thankYouNote.trim()
            val next = content.copy(isActive = false)
            _state.update {
                it.copy(
                    state = ManageTrainState.Loaded(next),
                    sheetMode = ManageTrainSheetMode.CLOSED,
                    toast = "Train closed · thanks sent to ${content.helpersValue} helpers",
                )
            }
            viewModelScope.launch {
                if (note.isNotEmpty()) {
                    repo.postUpdate(trainId, SupportTrainUpdateBody(body = note))
                }
                repo.complete(trainId)
            }
        }

        companion object {
            /** Nav-arg key for the train id. Keep in sync with [TRAIN_ID_KEY] in `ChildRoutes`. */
            const val TRAIN_ID_KEY: String = "supportTrainId"
        }
    }
